### PR TITLE
api: move route.service.strategy to route.loadBalancerPolicy

### LINKED ISSUE #1668: api,internal/dag: move route.service.strategy to route
As part of buying insurance for #1119, similar to #1630, move route.service.strategy to route.strategy. Bonus points for calling this `route.loadbalancingStrategy`

### PR DESCRIPTION
**Pull request by davecheney**
Updates #1119
Fixes #1668

This PR moves the per service lb strategy to a per route lb policy.
The lb policy type provides a place to hold the load balancer policy
strategy.

Signed-off-by: Dave Cheney <dave@cheney.net>

### CODE REVIEW

File: internal/featuretests/envoy.go [hunk 1]
```diff
@@ -102,6 +102,47 @@ func withWebsocket(route *envoy_api_v2_route.Route_Route) *envoy_api_v2_route.Ro
 	)
 	return route
 }
+func withSessionAffinity(route *envoy_api_v2_route.Route_Route) *envoy_api_v2_route.Route_Route {
```
Comment (jpeach) [inline internal/featuretests/envoy.go]: Nit: newline above here.

File: internal/featuretests/loadbalancerpolicy_test.go [hunk 2]
```diff
@@ -0,0 +1,249 @@
...
+	v2 "github.com/envoyproxy/go-control-plane/envoy/api/v2"
+	ingressroutev1 "github.com/projectcontour/contour/apis/contour/v1beta1"
+	projcontour "github.com/projectcontour/contour/apis/projectcontour/v1"
+	"github.com/projectcontour/contour/internal/envoy"
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/intstr"
+)
+
+// session affinity is only availavble in ingressroute and httpproxy
```
Comment (jpeach) [inline internal/featuretests/loadbalancerpolicy_test.go]: Nit: "available"

Comment (jpeach) [inline internal/featuretests/loadbalancerpolicy_test.go]: Nit: "available"

File: internal/dag/builder.go [hunk 3]
```diff
@@ -778,11 +778,11 @@ func (b *Builder) processIngressRoutes(sw *ObjectStatusWriter, ir *ingressroutev
 					}
 				}
 				r.Clusters = append(r.Clusters, &Cluster{
-					Upstream:             s,
-					LoadBalancerStrategy: service.Strategy,
-					Weight:               service.Weight,
-					HealthCheckPolicy:    ingressrouteHealthCheckPolicy(service.HealthCheck),
-					UpstreamValidation:   uv,
+					Upstream:           s,
+					LoadBalancerPolicy: service.Strategy,
```
[the remark this answers is not in the document]
Comment (davecheney) [reply internal/dag/builder.go]: not in this case. this is the legacy Ingressroute version.

File: internal/dag/builder.go [hunk 4]
```diff
@@ -878,8 +878,8 @@ func (b *Builder) processTCPProxy(sw *ObjectStatusWriter, ir *ingressroutev1.Ing
 				return
 			}
 			proxy.Clusters = append(proxy.Clusters, &Cluster{
-				Upstream:             s,
-				LoadBalancerStrategy: service.Strategy,
+				Upstream:           s,
+				LoadBalancerPolicy: service.Strategy,
```
[the remark this answers is not in the document]
Comment (davecheney) [reply internal/dag/builder.go]: as above

**Pull request merged**